server/src/main/java/org/tctalent/server/api/admin/UserAdminApi.java

```diff
@@ -135,12 +136,27 @@ public void sendVerifyEmailRequest(@RequestBody SendVerifyEmailRequest request)
     }
 
     @GetMapping("/verify-email/{token}")
-    public void verifyEmail(@PathVariable("token") String token) throws ExpiredEmailTokenException, InvalidEmailVerificationTokenException {
-        VerifyEmailRequest request = new VerifyEmailRequest();
-        request.setToken(token);
-        userService.verifyEmail(request);
+    public ModelAndView verifyEmail(@PathVariable("token") String token) {
+        ModelAndView modelAndView = new ModelAndView();
+        try {
+            VerifyEmailRequest request = new VerifyEmailRequest();
+            request.setToken(token);
+            userService.verifyEmail(request);
+            User user = userService.getUserByEmailVerificationToken(token);
+
+            modelAndView.setViewName("verify-email-success"); // src/main/resources/templates/verify-email-success.html
+            modelAndView.addObject("displayName", user.getDisplayName());
+        } catch (ExpiredEmailTokenException e) {
+            modelAndView.setViewName("verify-email-failure"); // src/main/resources/templates/verify-email-failure.html
+            modelAndView.addObject("displayName", "User");
```

Comment:
The purpose of the refactoring was so that you could replace this with:

```suggestion
            modelAndView.addObject("displayName", user.getDisplayName());
```

And also on line number 154.

Line 145 can be executed out of the try block (so can lines 142 and 143 - they don't really need to be in the try block.) 

That way the variable "user" is visible everywhere.

Reply:
Actually you will need to wrap the getUserByEmailVerificationToken with a try catch to detect an invalid token (not expired though) - which a different kind of error.

But with these changes you will be able to give more useful error messages for expired tokens - giving the user name.